src/Microsoft.VisualStudio.Threading/RarelyRemoveItemSet`1.cs

```diff
@@ -0,0 +1,242 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.Threading
+{
+    using System;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// A collection optimized for usually small number of elements, and items are rarely removed.
+    /// Note: this implementation is not thread-safe. It must be protected to prevent race conditions.
+    /// </summary>
+    /// <typeparam name="T">The type of elements to be stored.</typeparam>
+    internal struct RarelyRemoveItemSet<T>
+        where T : class
+    {
+        private const int MaxExpansionSize = 16 * 1024;
+
+        /// <summary>
+        /// The single value or array of values stored by this collection. Null if empty.
+        /// </summary>
+        private object? value;
+
+        /// <summary>
+        /// The number of items.
+        /// </summary>
+        private int count;
+
+        /// <summary>
+        /// Adds an element to the collection.
+        /// </summary>
+        public void Add(T value)
+        {
+            if (this.value is T[] valueArray)
+            {
+                if (valueArray.Length > this.count)
+                {
+                    valueArray[this.count] = value;
+                }
+                else
+                {
+                    int nextSize = valueArray.Length > MaxExpansionSize ? (valueArray.Length + MaxExpansionSize) : valueArray.Length * 2;
+
+                    Array.Resize(ref valueArray, nextSize);
+                    valueArray[this.count] = value;
+                    this.value = valueArray;
+                }
+            }
+            else
+            {
+                if (this.count == 0)
+                {
+                    this.value = value;
+                }
+                else
+                {
+                    Assumes.Equals(1, this.count);
+                    valueArray = new T[2] { (T)this.value!, value };
+                    this.value = valueArray;
+                }
+            }
+
+            this.count++;
+        }
+
+        /// <summary>
+        /// Removes an element from the collection.
+        /// </summary>
+        public void Remove(T value)
+        {
+            if (this.count == 0)
+            {
+                return;
+            }
+
+            if (this.value is T[] valueArray)
+            {
+                for (int i = 0; i < this.count; i++)
+                {
+                    if (valueArray[i] == value)
```

Comment:
Can this loop exit early when an item is found, or must it remove all values?

Reply:
good catch... apparently missed it.